pilot/pkg/networking/core/listener_waypoint.go

```diff
@@ -246,13 +247,26 @@ func (lb *ListenerBuilder) buildWaypointInternal(wls []model.WorkloadInfo, svcs
 					TCP:  match.ToChain(tcpClusterName),
 					HTTP: match.ToChain(httpClusterName),
 				}))
+				portProtocols[port.Port] = protocol.Unsupported
 			} else if port.Protocol.IsHTTP() {
 				// Otherwise, just insert HTTP/TCP
 				chains = append(chains, httpChain)
 				portMapper.Map[portString] = match.ToChain(httpChain.Name)
+				// TCP and HTTP on the same port, mark it as requiring sniffing
+				if portProtocols[port.Port] != "" && portProtocols[port.Port] != protocol.HTTP {
+					portProtocols[port.Port] = protocol.Unsupported
+				} else {
+					portProtocols[port.Port] = protocol.HTTP
+				}
 			} else {
 				chains = append(chains, tcpChain)
 				portMapper.Map[portString] = match.ToChain(tcpChain.Name)
+				// TCP and HTTP on the same port, mark it as requiring sniffing
+				if portProtocols[port.Port] != "" && portProtocols[port.Port] != protocol.TCP {
```

Comment:
nit: would it be clearer to have a closure like

```go
recordProtocol := func (p) {
  if proto, ok := portProtocols[p.Port]; ok && proto != p.Protocol {
  // unsupported
  }  else {
    // first instance of the port #, record the protocol
  }
}
```

Maybe that's awkward b/c we want to track the protocol _type_ (we may have different kinds of TCP and HTTP)